src/aiokatcp/client.py

```diff
@@ -923,177 +971,318 @@ def __init__(self, client: Client) -> None:
+        # Ensure that we're subscribed to the sensors we want to be.
+        add = []
+        remove = []
+        for s in self._need_subscribe:
+            if s.watchers and not s.subscribed:
+                add.append(s)
+            elif not s.watchers and s.subscribed:
+                remove.append(s)
+            else:  # pragma: nocover
+                assert False, "_need_subscribe inconsistent with sensor state"
+        # Sorting is not needed for correctness, but it makes behaviour
+        # predictable and easier to test.
+        if add:
+            add.sort(key=lambda s: s.info.name)
+            await self._set_sampling(add, "auto")
+        # Check _need_sensor_list again: if there are changes during adding we
+        # first re-read the sensor list rather than doing more work.
+        if not self._need_sensor_list and remove:
+            remove.sort(key=lambda s: s.info.name)
+            await self._set_sampling(remove, "none")
 
     def _connected(self) -> None:
-        self._sampling_set.clear()
-        self._trigger_update()
+        for s in self._sensors.values():
+            s.subscribed = False
+            self._update_need_subscribe(s)
+        self._need_sensor_list = True
+        self._trigger_update(True)
 
     def _disconnected(self) -> None:
-        self._sampling_set.clear()
-        self._cancel_update()
-        for watcher in self._watchers:
-            watcher.state_updated(SyncState.DISCONNECTED)
+        self._set_state(SyncState.DISCONNECTED)
+        self._update_event.clear()
 
     def _interface_changed(self, *args: bytes) -> None:
         # This could eventually be smarter and consult the args
+        self._need_sensor_list = True
         self._trigger_update()
 
     def _sensor_status(self, timestamp: core.Timestamp, n: int, *args) -> None:
         if len(args) != 3 * n:
             raise FailReply("Incorrect number of arguments")
+        # TODO: could limit _batch to only watchers that are affected
         with self._batch():
             for i in range(n):
                 name = "<unknown>"
                 try:
                     name = core.decode(str, args[3 * i])
                     status = core.decode(sensor.Sensor.Status, args[3 * i + 1])
                     value = args[3 * i + 2]
-                    for watcher in self._watchers:
+                    s = self._sensors.get(name)
+                    if s is None:
+                        self.logger.warning("Received update for sensor %s we haven't seen", name)
+                        continue
+                    s.reading = sensor.Reading(value=value, status=status, timestamp=timestamp)
+                    for watcher in s.watchers:
                         watcher.sensor_updated(name, value, status, timestamp)
+                    if not s.watchers and not s.subscribed:
+                        # We received a value, so we must be subscribed even if
+                        # didn't think we were.
```

Comment:
I think this needs to be done even if `s.watchers` is non-empty, to preserve the invariant that s.reading is not None => s.subscribed. But also need to make sure that isn't going to conflict with what happens in _set_sampling, since this code can run in the middle of that process.

Reply:
I believe it's safe even if it happens in the middle. There are a few cases:
1. We're subscribing in `_set_sampling`, and we get the initial sensor value. This is a normal case. We'll just be setting `subscribed` early, and `_set_sampling` will set it again.
2. We're unsubscribing in `_set_sampling`, and we got a sensor update just before the unsubscription occurred. Then `_set_sampling` will set it back to false.

This does depend on never receiving a "late" update after unsubscribing, which might cause us to think we were subscribed when we were not. Normally that's harmless since we'd just unsubscribe again, but if something else changed at the same time to make us want to subscribe (a new watcher), we'd skip re-subscribing.